sequelize/relationships

FelixPriori opened the pull request:
# In this PR
* Defines ClassProgress, AssignedBadge, Badge, Class relationships in `back-end/db/models/relationships.js`
* Creates a factory function for ClassProgress, AssignedBadge.
* Writes tests for ClassProgress, AssogmedBadge, Badge, and Class models
* Fixes some key names which were not matching db
* Renames AssignedBadges => AssignedBadge for consistency

FelixPriori commented:
# Notes

Added relationship for AssignedBadges, fixed import, and created its factory function.
Wrote a simple test, but I can't get it to pass ... second set of eyes would be appreciated.

```
 FAIL  tests/models/assignedBadges.test.js
  assigned badges model
    ✕ can fetch the associated badge (39ms)

  ● assigned badges model › can fetch the associated badge

    TypeError: Cannot read property 'getBadge' of null

      21 |   it("can fetch the associated badge", async () => {
      22 |     const badgeForAdventurer = await AssignedBadges.findByPk(assignedBadges.id);
    > 23 |     const adventurerBadge = await badgeForAdventurer.getBadge();
         |                                                      ^
      24 |     expect(adventurerBadge.id).toEqual(badge.id);
      25 |   });
      26 | });

      at Object.it (tests/models/assignedBadges.test.js:23:54)
```

FelixPriori commented:
# Notes

Fixed some of the association in `relationship` to be able to do the get calls, but the tests in `tests/models/assignedBadges.test.js` are still not passing. See error above.

cawel commented on back-end/tests/models/assignedBadges.test.js:
```diff
@@ -0,0 +1,31 @@
+  createAdventurer,
+  createBadge,
+} = require("../seeds");
+const AssignedBadges = require("../../db/models/classProgress");
```
All tests in that file are actually failing 😱 
That's because you're not importing from the right model file here 😉

FelixPriori replied:
# derp

cawel commented on back-end/tests/models/classProgress.test.js:
```diff
@@ -0,0 +1,38 @@
+    const progressClass = await progressForAdventurer.getClass();
+    expect(progressClass.id).toEqual(classInstance.id);
+  });
+});
```
That's a totally personal style, but I would add a blank line between each test example, so that it reads better. And so, in all the test files 😉

cawel commented on back-end/tests/models/assignedBadges.test.js:
```diff
@@ -0,0 +1,31 @@
+    badge = await createBadge();
+    assignedBadges = await createAssignedBadges({
+      badge_id: badge.id,
+      adventurer_id: adventurer.id,
```
There's indeed an `adventurer_id` column in that `assigned_badges` table, but the ERD in the readme says `user_id`. Might be worth updating the readme.

FelixPriori replied:
yes, I got confused by that as well because I was using the ERD as a reference. I will go change it today.

cawel replied:
I don't see the ERD update in that PR. Did you plan to do it in another PR?

FelixPriori replied:
The file itself has been updated, but it's a .png and it is name exactly the same so I don't know if there's a dif?

FelixPriori replied:
![image](https://user-images.githubusercontent.com/53023071/82846869-f6341180-9eb8-11ea-8a3c-702490d6641f.png)

cawel replied:
it's binary file, and since you made a change to it, quite naturally, there will be a diff 😉

FelixPriori commented:
# Notes:
* Reviewed code according to @cawel 's comments
* Changed name of `AssignedBadges` to `AssignedBadge` and refactored code to match this change.
* All tests are now passing.
* Added passing tests for `Badge` model

cawel commented on back-end/db/models/badges.js:
```diff
@@ -36,6 +36,6 @@ const Badge = sequelize.define(
 );
 
-// Badge.belongsToMany(User, { through: "AssignedBadges" });
+// Badge.belongsToMany(User, { through: "AssignedBadge" });
```
bye bye 
```suggestion
```

cawel commented on back-end/db/models/users.js:
```diff
@@ -44,6 +44,6 @@ const User = sequelize.define(
 );
 
-// User.belongsToMany(Badge, { through: "AssignedBadges" });
+// User.belongsToMany(Badge, { through: "AssignedBadge" });
```
bye bye
```suggestion
```

cawel commented:
The relationship "Quest belongs to Class" is missing.

cawel commented on back-end/db/models/relationships.js:
```diff
@@ -15,6 +15,11 @@ Quest.belongsTo(User, {
 
+Quest.belongsTo(Class, {
+  foreignKey: "class_id",
+  as: "class",
```
pretty sure the `as` part is unnecessary.
```suggestion
```

FelixPriori replied:
do we still want to refactor all that to match the default sequelize fk format?

cawel replied:
It would make things easier and terser if we were to respect sequelize's nomenclature . But I'd do that in another PR, after this one is merged.

FelixPriori commented:
# Done

This PR should be ready now!
